scripts/deployTokenLocks.js

```diff
@@ -0,0 +1,155 @@
+/* ./scripts/deployTokenLock.js */
+require('log-timestamp');
+const hre = require('hardhat');
+const toBN = hre.ethers.BigNumber.from;
+const fs = require('fs');
+
+const Q18 = toBN(10).pow(toBN(18));
+
+// @dev - group array by multiple keys - used to fold multiple lock requirements into
+// the same token lock contract
+function multipleGroupByArray (dataArray, groupPropertyArray) {
+  const groups = {};
+  dataArray.forEach(item => {
+    const group = JSON.stringify(groupPropertyArray(item));
+    groups[group] = groups[group] || [];
+    groups[group].push(item);
+  });
+  return Object.keys(groups).map(function (group) {
+    return groups[group];
+  });
+}
+
+// Load the deployment configuration file and set up constants and contract arguments
+const network = hre.hardhatArguments.network;
+console.log('Operating on network: ' + network);
+const actionsFilepath = './scripts/' + network + '/logs/actions.json';
+let lockConfig;
+try {
+  lockConfig = JSON.parse(fs.readFileSync('./scripts/' + network + '/deployTokenLock.config.json', 'utf-8'));
+} catch (err) {
+  console.error(err);
+  process.exit(1);
+}
+// Calculate Unix timestamp from UTC datetime
+const TGE_TIME = Math.floor(new Date(lockConfig.TGETime).getTime() / 1000);
+
+// Load actions object (TODO: move actions function to utils)
+let actions;
+try {
+  // load actions file
+  actions = JSON.parse(fs.readFileSync(actionsFilepath, 'utf-8'));
+} catch (err) {
+  if (err.code === 'ENOENT') {
+    console.log('TokenLock deployment requires an existing actions.json file to retrieve WFAIR contract');
+    process.exit(1);
+  } else {
+    console.error(err);
+    process.exit(1);
+  }
+};
+console.log('Actions will be logged to ' + actionsFilepath);
+
+// Retrieve WFAIR contract address from config file
+const WFAIR_CONTRACT = actions.token.address;
+
+// Add locks key if it doesn't exist
+if (!('locks' in actions)) { actions.locks = []; };
+
+// Create an array of arguments for the total list of lock contracts by grouping entries that
+// can be deployed to the same lock contract due to identical startDate, vesting period, cliff, and initial
+// payout values
+const lockGroups = multipleGroupByArray(lockConfig.lockRequests, function (item) {
+  return [item.vestingPeriod, item.cliffPeriod, item.initialReleaseFraction, item.delay];
+});
+console.log('Number of lock functions to deploy: ' + lockGroups.length);
+console.log(lockGroups);
+
+// TODO: check that each element of lockGroups doesn't contain the same address twice
+// as that would over-write the first lock with the second and lose tokens
+
+// Minimum ETH balance - to be determined from gas analysis
+const MIN_ETH = toBN('100000000000000000'); // TODO: should this be in the deployTokenLockConfig.json file?
+
+// Calculate total WFAIR supply requirement, check for cliff/initial conflicts,
+// and create arguments for TokenLock contract
+let TOTAL = toBN(0);
```

Comment:
Moved it to utils, but each tally uses slightly different key values of the parent object, because one is the initial dictionary, the second is an array, and the third is an analysis of the returned transaction which requires a key of an array of an array.

Reply:
I guess I could pass a property variable. Might do that later on.